File: java/src/main/java/org/libpetri/core/Instance.java
package org.libpetri.core;

import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;

/**
 * A typed module instance produced by {@link SubnetDef#instantiate(String, Object)},
 * per {@code spec/11-modular-composition.md} requirements <b>MOD-010</b>
 * (creation), <b>MOD-011</b> (typed handle map), <b>MOD-012</b>
 * (per-instance state isolation), and <b>MOD-030</b> (action binding).
 *
 * <p>An instance carries:
 * <ul>
 *   <li>The {@link #prefix()} string used to rename body elements
 *       (per [MOD-010] separator {@code "/"});</li>
 *   <li>A reference to the originating {@link SubnetDef};</li>
 *   <li>The renamed body — a structurally valid {@link PetriNet} per
 *       [CORE-040];</li>
 *   <li>Typed lookup handles for ports and channels keyed by their
 *       <b>original</b> (pre-prefix) names;</li>
 *   <li>The {@code params} value supplied at instantiation.</li>
 * </ul>
 *
 * <h2>Scaffolding status</h2>
 * <p>This class ships in scaffolding form: the constructor is
 * package-private and the rich {@link #bindActions(Map)} method throws
 * {@link UnsupportedOperationException} until the {@code instantiate} rename
 * pass lands in the next task. Direct accessors are wired up so downstream
 * code can take dependencies on the API shape today without waiting for the
 * rename pass.
 *
 * @param <P> parameter type (use {@link Void} for unparameterised subnets)
 */
public final class Instance<P> {

    private final String prefix;
    private final SubnetDef<P> def;
    private final PetriNet renamedBody;
    private final Map<String, Place<?>> portHandles;
    private final Map<String, Transition> channelHandles;
    private final P params;

    /**
     * Package-private constructor invoked by {@link SubnetDef#instantiate}.
     * Direct construction by user code is not permitted; the rename pass and
     * handle population are responsibilities of {@code SubnetDef}.
     */
    Instance(
        String prefix,
        SubnetDef<P> def,
        PetriNet renamedBody,
        Map<String, Place<?>> portHandles,
        Map<String, Transition> channelHandles,
        P params
    ) {
        this.prefix = Objects.requireNonNull(prefix, "prefix");
        this.def = Objects.requireNonNull(def, "def");
        this.renamedBody = Objects.requireNonNull(renamedBody, "renamedBody");
        this.portHandles = Map.copyOf(portHandles);
        this.channelHandles = Map.copyOf(channelHandles);
        this.params = params;
    }

    /** Returns the prefix string supplied at {@link SubnetDef#instantiate}. */
    public String prefix() {
        return prefix;
    }

    /** Returns the originating subnet definition. */
    public SubnetDef<P> def() {
        return def;
    }

    /** Returns the renamed body net (structurally valid {@link PetriNet}). */
    public PetriNet renamedBody() {
        return renamedBody;
    }

    /** Returns the parameter value supplied at instantiation, or {@code null} for unparameterised instances. */
    public P params() {
        return params;
    }

    /**
     * Returns the renamed {@link Place} corresponding to the named port.
     *
     * <p>The port name is the <b>original</b> (pre-prefix) name as declared
     * in the subnet's {@link Interface}. The {@code tokenType} parameter is
     * cross-checked against the underlying place's declared token type;
     * a mismatch raises {@link IllegalArgumentException} per [MOD-011].
     *
     * @throws IllegalArgumentException when the port name is unknown or
     *         when {@code tokenType} disagrees with the declared port type
     */
    @SuppressWarnings("unchecked")
    public <T> Place<T> port(String name, Class<T> tokenType) {
        var place = portHandles.get(name);
        if (place == null) {
            throw new IllegalArgumentException(
                "No port named '" + name + "' in instance '" + prefix + "'");
        }
        if (!place.tokenType().equals(tokenType)) {
            throw new IllegalArgumentException(
                "Port '" + name + "' has token type " + place.tokenType().getName()
                    + " but was requested as " + tokenType.getName());
        }
        return (Place<T>) place;
    }

    /**
     * Returns the renamed {@link Transition} corresponding to the named channel.
     *
     * @throws IllegalArgumentException when the channel name is unknown
     */
    public Transition channel(String name) {
        var t = channelHandles.get(name);
        if (t == null) {
            throw new IllegalArgumentException(
                "No channel named '" + name + "' in instance '" + prefix + "'");
        }
        return t;
    }

    /**
     * Returns the debug-UI descriptor for this instance per <b>MOD-041</b>.
     */
    public SubnetInstance descriptor() {
        return new SubnetInstance(
            prefix,
            def.name(),
            renamedBody.transitions(),
            Set.copyOf(portHandles.values()),
            params,
            Optional.empty()
        );
    }

    /**
     * Produces a derived instance whose specified transitions (named by their
     * <b>original</b>, pre-prefix names) carry the supplied actions per
     * <b>MOD-030</b>.
     *
     * <p>Mapping semantics:
     * <ul>
     *   <li>The keys of {@code actionsByOriginalName} are the
     *       <b>un-prefixed</b> transition names from the subnet definition's body.
     *       They are translated internally to the prefixed names of the renamed
     *       body before forwarding to {@link PetriNet#bindActions(Map)}.</li>
     *   <li>Transitions not named in the override map retain their shared
     *       default action (the action carried through from the body, by
     *       reference per [MOD-030]).</li>
     *   <li>Override is per-instance: invoking {@code bindActions} on this
     *       instance does not affect any other instance of the same
     *       {@link SubnetDef}.</li>
     *   <li>An override key that does not correspond to any transition in the
     *       subnet body is rejected with {@link IllegalArgumentException},
     *       naming the offending key.</li>
     * </ul>
     *
     * <p>The returned instance has the same {@link #prefix()}, {@link #def()},
     * {@link #portHandles port handles}, {@link #channelHandles channel
     * handles}, and {@link #params()} as this instance — only the action
     * bindings differ.
     *
     * <h3>Channel-identity invariant (compose contract)</h3>
     * <p>Implementations of this method (and any future {@link Instance}
     * derivation that returns a new {@code Instance} reusing the existing
     * {@code channelHandles} map) <b>MUST</b> preserve the identity — both
     * <i>name</i> and <i>arc structure</i> — of every transition referenced
     * by an {@link Interface} channel. The returned derived {@code Instance}
     * shares the same port and channel handle maps as {@code this}; those
     * handles must continue to resolve to a transition that is a member of
     * the derived {@link #renamedBody()} (lookup is by transition name).
     *
     * <p>If a refactor were to recompute channel-transition identity (for
     * example by renaming or re-keying the transitions while updating
     * actions), the channel-handle map would silently fall out of sync with
     * the renamed body and {@link PetriNet.Builder#compose(Instance, java.util.function.Consumer)}
     * would resolve channels to transitions that no longer exist in the
     * composed flat net. This invariant is checked at compose-time in
     * {@link PetriNet.Builder#compose(Instance, java.util.function.Consumer)};
     * a violation raises {@link IllegalStateException}.
     *
     * @param actionsByOriginalName map from original (un-prefixed) transition
     *                              name to the action to bind on this instance
     * @return a fresh {@link Instance} whose renamed body carries the override actions
     * @throws IllegalArgumentException when {@code actionsByOriginalName} contains
     *                                  a key that does not match any transition
     *                                  in the subnet body
     */
    public Instance<P> bindActions(Map<String, TransitionAction> actionsByOriginalName) {
        Objects.requireNonNull(actionsByOriginalName, "actionsByOriginalName");

        if (actionsByOriginalName.isEmpty()) {
            return this;
        }

        // Build the original-name set so we can validate keys cheaply.
        var bodyTransitionNames = new java.util.HashSet<String>(def.body().transitions().size());
        for (var t : def.body().transitions()) {
            bodyTransitionNames.add(t.name());
        }

        // Translate the override map keys: original-name -> prefixed-name.
        var prefixedActions = new java.util.LinkedHashMap<String, TransitionAction>();
        for (var entry : actionsByOriginalName.entrySet()) {
            var original = entry.getKey();
            if (!bodyTransitionNames.contains(original)) {
                throw new IllegalArgumentException(
                    "bindActions: no transition named '" + original
                        + "' in subnet '" + def.name() + "' (instance prefix '" + prefix + "')");
            }
            prefixedActions.put(prefix + "/" + original, entry.getValue());
        }

        // Delegate to the existing PetriNet.bindActions infrastructure (CORE-042).
        // The resolver leaves un-named transitions untouched (returning their
        // current action), per MOD-030 share-by-default semantics.
        var rebound = renamedBody.bindActions(name -> {
            var override = prefixedActions.get(name);
            if (override != null) return override;
            // Returning null tells PetriNet.bindActions to leave the existing action in place.
            return null;
        });

        return new Instance<>(prefix, def, rebound, portHandles, channelHandles, params);
    }
}

File: typescript/src/core/instance.ts
import type { PetriNet } from './petri-net.js';
import type { Place } from './place.js';
import type { SubnetDef } from './subnet-def.js';
import type { SubnetInstance } from './subnet-instance.js';
import type { Transition } from './transition.js';
import type { TransitionAction } from './transition-action.js';

/**
 * @internal Symbol key restricting construction to {@link SubnetDef.instantiate}
 * and the (future) `subnet-rewriter` module. Exported via the package-internal
 * factory {@link __createInstance} so the rewriter can construct Instances
 * without exposing the constructor publicly.
 */
const INSTANCE_KEY = Symbol('Instance.internal');

/**
 * A typed module instance produced by {@link SubnetDef.instantiate}, per
 * `spec/11-modular-composition.md` requirements **MOD-010** (creation),
 * **MOD-011** (typed handle map), **MOD-012** (per-instance state isolation),
 * and **MOD-030** (action binding).
 *
 * An instance carries:
 * - The {@link prefix} used to rename body elements (per [MOD-010] separator `"/"`);
 * - A reference to the originating {@link SubnetDef};
 * - The renamed body — a structurally valid `PetriNet` per [CORE-040];
 * - Typed lookup handles for ports and channels keyed by their **original**
 *   (pre-prefix) names;
 * - The `params` value supplied at instantiation.
 *
 * **Scaffolding status**: this class ships in scaffolding form. The
 * {@link bindActions} method throws `Error("not implemented")` until the
 * `instantiate` rename pass lands in the next task. Direct accessors are
 * wired up so downstream code can take dependencies on the API shape today.
 *
 * @typeParam P parameter type (use `void` for unparameterised subnets)
 */
export class Instance<P = void> {
  readonly prefix: string;
  readonly def: SubnetDef<P>;
  readonly renamedBody: PetriNet;
  readonly portHandles: ReadonlyMap<string, Place<unknown>>;
  readonly channelHandles: ReadonlyMap<string, Transition>;
  readonly params: P;

  /**
   * @internal Use {@link SubnetDef.instantiate} (or the internal factory
   * {@link __createInstance}) to create instances.
   */
  constructor(
    key: symbol,
    prefix: string,
    def: SubnetDef<P>,
    renamedBody: PetriNet,
    portHandles: ReadonlyMap<string, Place<unknown>>,
    channelHandles: ReadonlyMap<string, Transition>,
    params: P,
  ) {
    if (key !== INSTANCE_KEY) {
      throw new Error('Use SubnetDef.instantiate() to create Instance values');
    }
    this.prefix = prefix;
    this.def = def;
    this.renamedBody = renamedBody;
    this.portHandles = portHandles;
    this.channelHandles = channelHandles;
    this.params = params;
  }

  /**
   * Returns the renamed {@link Place} corresponding to the named port.
   *
   * The port name is the **original** (pre-prefix) name as declared in the
   * subnet's `Interface`. Per **MOD-022**, TypeScript enforces token-type
   * compatibility at compile time only; this method does not validate `T`
   * at runtime. A missing name raises an `Error`.
   *
   * @throws when the port name is unknown
   */
  port<T>(name: string): Place<T> {
    const p = this.portHandles.get(name);
    if (p === undefined) {
      throw new Error(`No port named '${name}' in instance '${this.prefix}'`);
    }
    return p as Place<T>;
  }

  /**
   * Returns the renamed {@link Transition} corresponding to the named channel.
   *
   * @throws when the channel name is unknown
   */
  channel(name: string): Transition {
    const t = this.channelHandles.get(name);
    if (t === undefined) {
      throw new Error(`No channel named '${name}' in instance '${this.prefix}'`);
    }
    return t;
  }

  /**
   * Returns the debug-UI descriptor for this instance per **MOD-041**.
   */
  descriptor(): SubnetInstance {
    const transitions: string[] = [];
    for (const t of this.renamedBody.transitions) transitions.push(t.name);
    const exposedPlaces: string[] = [];
    for (const p of this.portHandles.values()) exposedPlaces.push(p.name);
    return {
      prefix: this.prefix,
      defName: this.def.name,
      transitions,
      exposedPlaces,
      params: this.params,
      parentPrefix: null,
    };
  }

  /**
   * Produces a derived instance whose specified transitions (named by their
   * **original**, pre-prefix names) carry the supplied actions per **MOD-030**.
   *
   * For each entry `[originalName, action]`, the renamed body is searched for
   * the transition whose name equals `prefix + "/" + originalName`. The
   * resulting instance shares the original `def`, `prefix`, `params`, and
   * port/channel handle topology — only the renamed body is rebuilt with new
   * actions. Per **MOD-030**, calling `bindActions` on one instance does NOT
   * affect the actions held by other instances of the same `def`.
   *
   * Unrecognised original names raise an `Error` so typos surface eagerly.
   */
  bindActions(actionsByOriginalName: Record<string, TransitionAction>): Instance<P> {
    // Build a Map<prefixedName, TransitionAction> for the resolver.
    // Validate every supplied original name against the renamed body's
    // transition set to surface typos eagerly.
    const prefixedByName = new Map<string, TransitionAction>();
    const renamedNames = new Set<string>();
    for (const t of this.renamedBody.transitions) {
      renamedNames.add(t.name);
    }

    for (const originalName of Object.keys(actionsByOriginalName)) {
      const prefixed = this.prefix + '/' + originalName;
      if (!renamedNames.has(prefixed)) {
        throw new Error(
          `Instance.bindActions: no transition '${originalName}' (resolved as ` +
          `'${prefixed}') in instance '${this.prefix}' of subnet '${this.def.name}'`,
        );
      }
      prefixedByName.set(prefixed, actionsByOriginalName[originalName]!);
    }

    // Use the existing PetriNet.bindActionsWithResolver — it preserves the
    // existing action when the resolver returns it (see petri-net.ts), so
    // unaffected transitions land in the new net by reference (sharing per
    // MOD-030).
    const reboundBody = this.renamedBody.bindActionsWithResolver((name) => {
      const action = prefixedByName.get(name);
      if (action !== undefined) return action;
      // Return the existing action so PetriNet.bindActionsWithResolver
      // short-circuits the rebuild for unaffected transitions.
      for (const t of this.renamedBody.transitions) {
        if (t.name === name) return t.action;
      }
      // Unreachable — the resolver is only called for transitions in this.renamedBody.
      /* istanbul ignore next */
      throw new Error(`Instance.bindActions: resolver invoked with unknown name '${name}'`);
    });

    // Rebuild the channel handles against the rebound body so callers see
    // the post-rebind transition (with its new action) when they ask for a
    // channel by name.
    const reboundChannelHandles = new Map<string, Transition>();
    if (this.channelHandles.size > 0) {
      const byName = new Map<string, Transition>();
      for (const t of reboundBody.transitions) {
        byName.set(t.name, t);
      }
      for (const [name, oldT] of this.channelHandles) {
        const refreshed = byName.get(oldT.name);
        if (refreshed === undefined) {
          /* istanbul ignore next */
          throw new Error(
            `Instance.bindActions: channel '${name}' transition '${oldT.name}' not found in rebound body`,
          );
        }
        reboundChannelHandles.set(name, refreshed);
      }
    }

    return __createInstance<P>(
      this.prefix,
      this.def,
      reboundBody,
      this.portHandles,
      reboundChannelHandles,
      this.params,
    );
  }
}

/**
 * @internal Package-internal factory used by {@link SubnetDef.instantiate}
 * and the (future) `subnet-rewriter` module to construct {@link Instance}
 * values without re-exporting the Symbol-guarded constructor key publicly.
 *
 * NOT part of the public API surface — do NOT re-export from `core/index.ts`.
 */
export function __createInstance<P>(
  prefix: string,
  def: SubnetDef<P>,
  renamedBody: PetriNet,
  portHandles: ReadonlyMap<string, Place<unknown>>,
  channelHandles: ReadonlyMap<string, Transition>,
  params: P,
): Instance<P> {
  return new Instance<P>(
    INSTANCE_KEY,
    prefix,
    def,
    renamedBody,
    portHandles,
    channelHandles,
    params,
  );
}
